module/polls

timsmd opened the pull request:
closes  #39

created poll command(module), added to index.js. At this point is understands the poll create syntax and creates a poll object, sending in back as a message

timsmd committed:
created poll command module

created command, added to index.js. At this point is understands the poll create syntax and creates a poll object, sending in back as a message

timsmd committed:
styles fix

minor style fixes, indentation, comments etc.

timsmd committed:
yet another style fix

timsmd committed:
quotes are stripped from the answers added

fenruga commented on src/modules/poll.command.js:
```diff
@@ -0,0 +1,30 @@
+    poll_obj.question = input.substring(input.indexOf(' ') + 1 , input.indexOf('?'));
+
+    // the rest of string in split into answers, spaces inside quotes are ignored, quotes are then stripped
+    poll_obj.answers = input.substring(input.indexOf('?') + 1).replace(/\s+(?=([^"]*"[^"]*")*[^"]*$)/g, '|').replace(/['"]/g, "").split('|');
```
see_no_evil.jpg

fenruga commented on src/modules/poll.command.js:
```diff
@@ -0,0 +1,30 @@
+    poll_obj = {};
+
+    // separating the question by question mark
+    poll_obj.question = input.substring(input.indexOf(' ') + 1 , input.indexOf('?'));
```
Semi-offtop: it restrict us to poll form, btw. I mean, we always forced to formulate poll with question mark.
Also, look for `response.args` inside `parseCommand` module, maybe we need to update it to fit this case (with quotes). And it is solving the issue with `?`. If we don't mind this format, of course:
`/poll "some question/suggestion here" variant1 "variant2 with spaces"`

timsmd committed:
Merge branch 'develop' into timsmd

timsmd committed:
changes single args into parsed and raw

timsmd committed:
styles fix (with help of lint)

timsmd committed:
package lock update

timsmd committed:
Merge branch 'develop' into timsmd

timsmd committed:
Merge branch 'develop' into timsmd

timsmd committed:
Merge branch 'develop' into timsmd

timsmd committed:
command now ignores spaces inside quotes when splitting args

timsmd committed:
index typo fix

timsmd committed:
created in memory objects, some todos completed, comments fixed

fenruga commented:
lint

fenruga commented on src/modules/poll.command.js:
```diff
@@ -0,0 +1,71 @@
+
+const command = require('./command');
+
+let polls = {
```
looks like testdata >_>

fenruga commented on src/modules/poll.command.js:
```diff
@@ -0,0 +1,71 @@
+
+        // TODO connect to the database and get real generated ID
+        pollObj.pollId = Math.floor(Math.random() * 100);
+        response.output = JSON.stringify(pollObj, null, 2);
```
Let's just say something like "Poll with id ${pollId} successfully created"
With i18n, sure

fenruga commented on src/modules/poll.command.js:
```diff
@@ -0,0 +1,71 @@
+
+    } else if (response.rawArgs.length == 1) {
+        let pollId = response.rawArgs[0];
+        response.output = JSON.stringify(getPolls(pollId), null, 2);
```
and here

fenruga commented on src/modules/poll.command.js:
```diff
@@ -0,0 +1,71 @@
+        response.output = JSON.stringify(getPolls(pollId), null, 2);
+    }  else {
+        console.log(getPolls());
+        response.output = JSON.stringify(getPolls(), null, 2);
```
and here

timsmd committed:
index.js lint fixes

timsmd committed:
style fixes (lint)

timsmd committed:
polls response output implemented (placeholders for votes for now)

timsmd committed:
style fixes (lint)

timsmd committed:
getPolls->getPoll renamed, iterator fixed

timsmd committed:
vote and close commands templates created, configured in index

timsmd committed:
uncommented close in index

timsmd committed:
implemented some of the i18n logic

timsmd committed:
style fixes lint

timsmd commented:
Still have questions about using i18n while iterating an unknown amount of answers in response, when displaying info about a poll

fenruga commented on src/modules/command.js:
```diff
@@ -3,7 +3,7 @@ const { PREFIX } = require('../config');
 module.exports = rawCommand => (response, { input }) => {
     const [command, ...argData] = rawCommand.split(' ');
-    const [cmd, ...rawArgs] = input.split(' ');
+    const [cmd, ...rawArgs] = input.trim().replace(/\s+(?=([^"]*"[^"]*")*[^"]*$)/g, '|').replace(/['"]/g, '').split('|');
```
I want tests on it.
And maybe it's better to keep regexp patterns as `const` values with some meaningful names

timsmd committed:
more i18n logic implemented

need comments on using i18n in for loop

timsmd committed:
Merge branch 'develop' into timsmd

fenruga commented on i18n/en/common.csv:
```diff
@@ -9,3 +9,7 @@ ping, "pong4"
+,
+pollCreated, "Poll created successfully, vote by typing: /vote ${pollId} your_answer"
+pollInfo, "${day}/${month} ${question} | ${votes} votes | ID${pollId}\n"
+answerInfo, "${_poll.answers[i]} (${percentage}%)"
```
@dcversus that's what I'm talking about :) https://github.com/hyperloot/bot.hyperloot.net/pull/58#discussion_r243537885
@timsmd it will not work, you need to operate with exact value for now <__>

dcversus replied:
u a right! need expose to use templator for i18n and support cases like this))) i do it, or u?

fenruga replied:
I don't think we actually need it.
I mean, why should we, if we can pass value as is?
And in this actual case we even have `i` inside it
I suggest to switch pattern, to make it less confusing. Or smthing <__>

dcversus replied:
hmm, like simple mustashee? `{{string}}` hmmm....  need discuss it))

fenruga commented on src/index.js:
```diff
@@ -30,14 +33,19 @@ instance.use([
 
-    pong,
-    status,
+    [
```
why it is grouped in subchain?

fenruga commented on src/modules/poll.command.js:
```diff
@@ -0,0 +1,108 @@
+    return polls;
+};
+
+const getPollStringInfo = function (_poll, { i18n }) {
```
No need to be destructive here -- just pass and get `i18n` as it is
And again: why `_` ?

fenruga commented on src/modules/poll.command.js:
```diff
@@ -0,0 +1,108 @@
+    if (id in polls) {
+        return polls[id];
+    }
+    return polls;
```
I mean...it should be described in some way.
I will be very surprised, if I write `getPoll(666)` and will recieve all exicting polls
Imo it is better to say, that we have no such poll

fenruga commented on src/modules/poll.command.js:
```diff
@@ -0,0 +1,108 @@
+    const day = _poll.dateCreated.getDate();
+    const month = _poll.dateCreated.getMonth();
+    const question = _poll.question;
+    // TODO get real votes count from db
```
You can start your work with db now

dcversus replied:
I give example from my module later :)

dcversus commented on src/index.js:
```diff
@@ -21,6 +21,9 @@ const updateExp = require('./modules/updateExp');
 const pong = require('./modules/pong.command');
+const poll = require('./modules/poll.command');
+const vote = require('./modules/vote.command');
+const close = require('./modules/close.command');
```
expose ONE poll module, and inside keep all three commands... and keep close/vote/poll inside "poll" folder, ok?

timsmd replied:
OK

dcversus commented on src/modules/poll.command.js:
```diff
@@ -0,0 +1,108 @@
+        const pollObj = {};
+
+        // raw arguments include question and an unknown ammount of answers, which are read into variables
+        [pollObj.question, ...pollObj.answers] = response.rawArgs;
```
command module provide spread functionality now, use it

fenruga replied:
didn't get it

dcversus replied:
no points to use `rawArgs`... now u can use args and something like `command('polls question ...answers')`

dcversus commented on src/modules/poll.command.js:
```diff
@@ -0,0 +1,108 @@
+
+        response.output = i18n('pollCreated', { pollId });
+        // response.output = `Poll created successfully, vote by typing: /vote ${pollObj.pollId} your_answer`;
+    } else if (response.rawArgs.length === 1) {
```
dont use else if or something like this... in most cases u can do several `if` and several `return` in code

fenruga replied:
It is my privilege to use it XD

dcversus commented on src/modules/poll.command.js:
```diff
@@ -0,0 +1,108 @@
+        response.output = getPollStringInfo(_poll, { i18n });
+    } else {
+        // getting info on all the existing polls
+        Object
```
fuuu dirty(((

use object with [id]: { ... } and array with ids and then easy 
```
const polls = {
  1: {...},
  2: {...},
  3: {...},
  ...
};

const polls_ids = [1,2,3, ...];

response.output = i18n('pollInfoHeader', { user });
response.output += polls_ids.map(id => i18n('pollInfo', polls[id]));
```

dcversus reviewed (changes requested):
good morning

timsmd committed:
Merge branch 'develop' into timsmd

timsmd committed:
remove the extra commands

timsmd committed:
added i18n for poll

timsmd committed:
some poll command refactoring

timsmd committed:
disabled autoreactions

dcversus commented on src/index.js:
```diff
@@ -30,7 +30,7 @@ instance.use([
         event('message'),
         addExp(1),
-        autoReaction,
+        // autoReaction,
```
why u push it?

dcversus replied:
please uncomment it, it need :)

dcversus commented on src/modules/poll.command.js:
```diff
@@ -61,48 +67,52 @@ const getPollStringInfo = function (_poll, { i18n }) {
+    updateModuleData('poll', {
+        list: [...list, newPoll],
+    })
+    console.log(newPoll)
```
dont commit consolelogs?

timsmd committed:
Merge branch 'develop' into timsmd

timsmd committed:
enabled poll, fixed the extra array issue

timsmd committed:
deleted test data

timsmd committed:
yet another fix on array of used modules

timsmd committed:
disabled autoreactions

timsmd committed:
refactoring

timsmd committed:
lint

timsmd committed:
quiz en i18n placeholder (same as russian)

timsmd committed:
i18n rework for polls + destructuring

timsmd committed:
minor fixes

timsmd committed:
changed the discord output test back to how it was

timsmd committed:
todos and test data improved

timsmd committed:
close command template + i18n

timsmd committed:
minor changes, i18n fixed for closing poll

timsmd committed:
changed variables names to more suitable

timsmd committed:
vote function template created

timsmd committed:
i18n for votes

timsmd committed:
lint error fixes + other minor fixes

timsmd committed:
more fixes to the god of fixes

timsmd committed:
comments and todos improved

timsmd committed:
lint (no broadcast usage)

timsmd committed:
flexible status display (i18n)

dcversus commented on src/modules/poll/index.js:
```diff
@@ -1,29 +1,41 @@
+    isOpen,
 }) {
+    let i18nOption = 'poll.header';
+    if (isOpen === false) i18nOption += '.wstatus';
```
its very hard to understand... please dont concate CONSTANTS (i18n selector is constants)

dcversus replied:
and prefer to line break and use curly braces in single if statment (for better readable)

dcversus commented on i18n/en/poll.csv:
```diff
@@ -1,6 +1,7 @@
 poll.alert, "Hey everyone @here, a new poll just popped up! vote by typing /vote your_answer"
-poll.header, "${day}/${month} ${question} | ${votes} votes | ID${pollId}\n"
+poll.header, "${day}/${month} ${question} | ${votes} votes | ID: ${pollId}\n"
+poll.header.wstatus, "${day}/${month} ${question} | ${votes} votes | ID: ${pollId} (${status})\n"
```
what doesnt mean "wstatus" here? what u want here? and why, one poll line named "header"... its like a "details"

dcversus commented on src/modules/poll/index.js:
```diff
@@ -1,29 +1,41 @@
+        // status,
     });
 
     for (let i = 0; i < options.length; i++) {
```
feee... use options.forEach and BETTER use

```
output += options.map(option => i18n('poll.option', option))
```

dcversus commented on src/modules/poll/index.js:
```diff
@@ -59,16 +71,18 @@ const addPoll = async function (response, {
+ * getPollById() implements the poll info logic
+ * the function gets info about a poll by it's mneumonic id
+ * usage /poll ID
+ */
```
dont need any descr, prefer write tests

fenruga replied:
tests > docs > comments > nothing

timsmd committed:
i18n fixes

timsmd committed:
module refactoring and rework

timsmd committed:
minor improvements

timsmd committed:
Merge branch 'develop' into timsmd

timsmd committed:
fixed error

timsmd committed:
poll i18n update

timsmd committed:
poll i18n update

timsmd committed:
db collections renamed

timsmd committed:
get poll rework in progress

dcversus commented on src/modules/poll/index.js:
```diff
@@ -62,7 +62,7 @@ const addPoll = async function (response, {
     };
 
-    updateModuleData('poll', {
+    updateModuleData('poll.polls', {
```
DONT DO IT!

```
updateModuleData(moduleName, newData)
```

dcversus commented on src/modules/poll/index.js:
```diff
@@ -137,7 +134,7 @@ const vote = async function (response, {
     updateModuleData,
 }) {
-    const { list = [] } = await getModuleData('vote');
+    const { list = [] } = await getModuleData('poll.votes');
```
not

dcversus commented on src/modules/poll/index.js:
```diff
@@ -146,7 +143,7 @@ const vote = async function (response, {
         dateVoted: new Date(),
     };
-    updateModuleData('vote', {
+    updateModuleData('poll.votes', {
```
dont like this

fenruga replied:
`updateModuleData('poll', someShit)`

timsmd committed:
i18n csv update

timsmd committed:
addPoll rework

timsmd committed:
getPollStringInfo deleted

timsmd committed:
pollsList rework

timsmd committed:
closePoll rework

timsmd committed:
i18n update

timsmd committed:
deleted descriptions

timsmd committed:
getPollbyId rework

timsmd committed:
pollsList update

timsmd committed:
Merge branch 'develop' into timsmd

timsmd committed:
installed human-readable-ids

timsmd committed:
minor i18n improvement

timsmd committed:
more rework on i18n

timsmd committed:
poll module overall improvement

timsmd committed:
Merge branch 'develop' into timsmd

merged develop into timsmd

timsmd committed:
i18n updates

timsmd committed:
refactoring

timsmd committed:
poll module logic improvements

timsmd committed:
one query instead of two

timsmd committed:
satisfying the lint

timsmd committed:
removed todo cause its actually done!

timsmd committed:
cant't vote for closed polls

timsmd committed:
one user one vote feature implemented

timsmd commented:
still have to implement vote by text or id feature, and that should do it

dcversus commented on i18n/en/common.csv:
```diff
@@ -9,3 +9,7 @@ ping, "pong4"
+,
+pollCreated, "Poll created successfully, vote by typing: /vote ${pollId} your_answer"
+pollInfo, "${day}/${month} ${question} | ${votes} votes | ID${pollId}\n"
+answerInfo, "${_poll.answers[i]} (${percentage}%)"
```
why it here? move it to poll.csv

dcversus commented on src/index.js:
```diff
@@ -18,14 +18,15 @@ const App = require('./app');
 const logText = require('./modules/logText');
 const updateExp = require('./modules/updateExp');
-const autoReaction = require('./modules/autoReaction');
+// const autoReaction = require('./modules/autoReaction');
```
UNCOMMENT it please.... dont touch(((

dcversus commented on src/modules/poll/index.js:
```diff
@@ -0,0 +1,197 @@
+    } = poll;
+
+    const day = dateCreated.getDate();
+    const month = dateCreated.getMonth() + 1;
```
dirty, please format it using http://momentjs.com/

dcversus reviewed (changes requested):
- [ ] add `ru` locale
- [ ] add vote by simple option input

timsmd committed:
reenabled autoreaction

timsmd committed:
removed old i18n data

timsmd committed:
enabled autoreaction for real this time

timsmd committed:
installed moment.js

timsmd committed:
moment js implemented

timsmd committed:
i18n updated to work with moment js

timsmd commented:
today's list:
- [x] implement moment js
- [x] fix can vote for anything bug
- [x] vote must understand both id and text
- [x] votes through messages monitoring
- [x] create ru locale .csv

timsmd committed:
i18n improved

timsmd committed:
fixed formatting

timsmd committed:
vote rework + id/text voting feature added

timsmd committed:
minor i18n fix

timsmd committed:
lint

timsmd committed:
checkVote template created

timsmd committed:
added newlines

not sure if its the best way, but .join() doesnt seem to work with i18n returned value

timsmd committed:
lint fixes

timsmd committed:
minor fix vote by id

timsmd committed:
voting by simple messages implemented

dcversus commented on src/modules/poll/index.js:
```diff
@@ -0,0 +1,303 @@
+    }
+
+    pollsList.filter(p => p.isOpen).forEach((p) => {
+        if (votesList.find(v => (v.pollId === p.pollId && v.voterId === id))) {
```
v? p? u a kidding?

please give normal names to variables!

fenruga replied:
semi-offtop: are `k`, `v` (key, value) okay?

dcversus commented on src/modules/poll/index.js:
```diff
@@ -0,0 +1,303 @@
+
+    pollsList.filter(p => p.isOpen).forEach((p) => {
+        if (votesList.find(v => (v.pollId === p.pollId && v.voterId === id))) {
+            return response;
```
return response from () => {} ??? what u want? 

need simple return to prevent below code exec? or u want prevent forEach?

timsmd replied:
I want to skip to the next iteration if the condition is true
same in the following if's inside this function

dcversus commented on src/modules/poll/index.js:
```diff
@@ -0,0 +1,303 @@
+                pollId: p.pollId,
+                option: p.options.indexOf(input),
+                dateVoted: new Date(),
+            };
```
line break?

dcversus commented on src/modules/poll/index.js:
```diff
@@ -0,0 +1,303 @@
+            };
+            updateModuleData('poll', {
+                votesList: [...votesList, newVote],
+            });
```
line break?

dcversus commented on src/modules/poll/index.js:
```diff
@@ -0,0 +1,303 @@
+                votesList: [...votesList, newVote],
+            });
+            const optionText = input;
+            const requestedPollId = p.pollId;
```
line break?

timsmd committed:
checking for existing polls

dcversus commented on src/modules/poll/index.js:
```diff
@@ -0,0 +1,303 @@
+                id,
+                requestedPollId,
+                optionText,
+            });
```
line break?

dcversus commented on src/modules/poll/index.js:
```diff
@@ -0,0 +1,303 @@
+            return response;
+        }
+
+        if (input - 1 >= 0 && input - 1 < p.options.length) {
```
input - 1 >= 0 WAT?

dcversus replied:
const variandIndex = input - 1;

timsmd committed:
styles fixes and return fixes

fenruga commented on i18n/en/poll.csv:
```diff
@@ -0,0 +1,17 @@
+poll.created, "Poll created successfully, here's the ID for results tracking and other stuff: ${pollId}"
```
add 'ru' too, please

timsmd replied:
i'm on it

timsmd commented:
- [x] create ticket for bug in polls
- [x] rename variables

fenruga commented on i18n/en/poll.csv:
```diff
@@ -0,0 +1,17 @@
+poll.created, "Poll created successfully, here's the ID for results tracking and other stuff: ${pollId}"
+poll.alert, "Hey everyone @here, a new poll just popped up! vote by typing /vote your_answer"
+
+poll.notfound, "I couldn't find a poll with this ID (${requestedPollId}), sorry, man("
```
go camelCase

timsmd committed:
minor fix

timsmd committed:
i18n switched to camelCase

fenruga commented on src/modules/poll/index.js:
```diff
@@ -0,0 +1,317 @@
+        dateCreated: new Date(),
+    };
+
+    updateModuleData('poll', {
```
Offtop: Imo, it was easier to make it as an object.
Now you have:
```js
poll: {
  pollList: [
    {pollId: id, ...someData},
    {pollId: otherId, ...otherData}
  ]
}
```
My option:
```js
poll: {
  pollList: {
    id: {...someData},
    otherId: {...otherData}
  }
}
```

fenruga replied:
No need to rework

timsmd committed:
refactoring

timsmd committed:
added russian poll.csv

fenruga commented on i18n/ru/poll.csv:
```diff
@@ -0,0 +1,17 @@
+poll.list, "Вот список всех открытых вопросов, все ка ты просил))) :\n"
+poll.header, "${date} ${question} | ${votesCount} всего голосов | ID: ${pollId}\n"
+poll.line, "${option} | ${optionVotes} голосов | (${percentage}%) |\n"
+poll.none, "В данный момент нет открытых голосований, может ты создашь новое сэмпай?.."
```
запятая, после "новое" XD

fenruga commented on src/modules/poll/index.js:
```diff
@@ -0,0 +1,319 @@
+        return response;
+    }
+
+    if (!currentPoll) {
```
Lots of copypasting
It means, that you should add subchain element, checking out different stuff, imo

fenruga replied:
Add it to next issue about this, please

fenruga commented on src/modules/poll/index.js:
```diff
@@ -0,0 +1,319 @@
+        };
+
+        updateModuleData('poll', {
+            votesList: [...votesList, newVote],
```
And there: imo it is easier to group it as an object, not list
fine for now

fenruga commented on src/modules/poll/index.js:
```diff
@@ -0,0 +1,319 @@
+
+        const inputIndex = input - 1;
+
+        if (inputIndex >= 0 && inputIndex < poll.options.length) {
```
copypaste, again
add to the issue to rid of it

timsmd committed:
russian grammar fixes
